File: app/src/main/java/com/jon/cotgenerator/utils/InputValidator.java
package com.jon.cotgenerator.utils;

import android.os.AsyncTask;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.concurrent.ExecutionException;
import java.util.regex.Pattern;

public class InputValidator {
    private InputValidator() { }

    public static boolean validateInt(final String str, final Integer min, final Integer max) {
        try {
            int number = Integer.parseInt(str);
            return (min == null || number >= min) && (max == null || number <= max);
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    public static boolean validateDouble(final String str, final Double min, final Double max) {
        try {
            double number = Double.parseDouble(str);
            return (min == null || number >= min) && (max == null || number <= max);
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    public static boolean validateString(final String str) {
        return (str != null && !str.isEmpty());
    }

    public static boolean validateString(final String str, final String regexPattern) {
        if (!validateString(str)) return false;
        return Pattern.compile(regexPattern).matcher(str).find();
    }

    public static boolean validateCallsign(String callsign) {
        /* These characters break TAK parsing when using XML, but not protobuf. I'll block them from both just to be safe */
        final Character[] disallowedCharacters = new Character[] {
                '<', '>',
        };
        for (Character character : disallowedCharacters) {
            if (callsign.indexOf(character) != -1) return false;
        }
        return true;
    }

    /* We don't care about the call to InetAddress.getByName() returning anything, all we want is to catch any exceptions from calling it */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    private static class ValidateHostnameTask extends AsyncTask<String, Void, Boolean> {
        @Override
        protected Boolean doInBackground(String... params) {
            try {
                InetAddress.getByName(params[0]);
                return true;
            } catch (UnknownHostException e) {
                return false;
            }
        }
    }

    public static boolean validateHostname(final String host) {
        try {
            return new ValidateHostnameTask().execute(host).get();
        } catch (InterruptedException | ExecutionException e) {
            return false;
        }
    }
}

File: common/src/main/java/com/jon/common/utils/InputValidator.kt
package com.jon.common.utils

import android.os.AsyncTask
import java.net.InetAddress
import java.net.UnknownHostException
import java.util.regex.Pattern

class InputValidator {
    fun validateInt(str: String, min: Int? = null, max: Int? = null): Boolean {
        return try {
            isInRange(str.toInt(), min, max)
        } catch (e: NumberFormatException) {
            false
        }
    }

    fun validateDouble(str: String, min: Double? = null, max: Double? = null): Boolean {
        return try {
            isInRange(str.toDouble(), min, max)
        } catch (e: NumberFormatException) {
            false
        }
    }

    fun validateString(str: String?): Boolean {
        return str != null && str.isNotEmpty()
    }

    fun validateString(str: String, regexPattern: String): Boolean {
        return if (!validateString(str)) false else Pattern.compile(regexPattern).matcher(str).find()
    }

    fun validateCallsign(callsign: String): Boolean {
        /* These characters break TAK parsing when using XML, but not protobuf. I'll block them from both just to be safe */
        val disallowedCharacters = arrayOf(
                '<', '>', '&', '\"'
        )
        return disallowedCharacters.none { callsign.contains(it) }
    }

    fun validateHostname(host: String): Boolean {
        return try {
            ValidateHostnameTask().execute(host).get()
        } catch (e: Exception) {
            false
        }
    }

    /* We don't care about the call to InetAddress.getByName() returning anything, all we want is to catch any exceptions from calling it */
    private class ValidateHostnameTask : AsyncTask<String, Void, Boolean>() {
        override fun doInBackground(vararg params: String): Boolean {
            return try {
                InetAddress.getByName(params[0])
                true
            } catch (e: UnknownHostException) {
                false
            }
        }
    }

    private fun <T : Comparable<T>> isInRange(value: T, min: T?, max: T?): Boolean {
        if (min != null && value < min) return false
        if (max != null && value > max) return false
        return true
    }
}
